src/main/scala/io/ino/solrs/AsyncSolrClient.scala

```diff
@@ -156,49 +161,80 @@ class AsyncSolrClient private (val loadBalancer: LoadBalancer,
     }
   }
 
-  def query(q: SolrQuery): Future[QueryResponse] = {
-    loadBalanceQuery(QueryContext(q))
+  def query(q: SolrQuery): F[QueryResponse] = {
+    futureFactory.toBase(
+    loadBalanceQuery(QueryContext(q)))
   }
 
   private def loadBalanceQuery(queryContext: QueryContext): Future[QueryResponse] = {
     loadBalancer.solrServer(queryContext.q) match {
-      case Some(solrServer) => queryWithRetries(solrServer, queryContext)
+      case Some(solrServer) =>
+        queryWithRetries(solrServer, queryContext)
       case None =>
+        val promise = futureFactory.newPromise[QueryResponse]
         val msg =
           if(queryContext.failedRequests.isEmpty) "No solr server available."
           else s"No next solr server available. These requests failed:\n- ${queryContext.failedRequests.mkString("\n- ")}"
-        Future.failed(new SolrServerException(msg))
+        promise.failure(new SolrServerException(msg))
+        promise.future
     }
   }
 
+  
   private def queryWithRetries(server: SolrServer, queryContext: QueryContext): Future[QueryResponse] = {
-    import scala.concurrent.ExecutionContext.Implicits.global
     val start = System.currentTimeMillis()
-    query(server, queryContext.q).recoverWith { case NonFatal(e) =>
-      val updatedContext = queryContext.failedRequest(server, (System.currentTimeMillis() - start) millis, e)
-      retryPolicy.shouldRetry(e, server, updatedContext, loadBalancer) match {
-        case RetryServer(s) =>
-          logger.warn(s"Query failed for server $server, trying next server $s. Exception was: $e")
-          queryWithRetries(s, updatedContext)
-        case StandardRetryDecision(Result.Retry) =>
-          logger.warn(s"Query failed for server $server, trying to get another server from loadBalancer for retry. Exception was: $e")
-          loadBalanceQuery(updatedContext)
-        case StandardRetryDecision(Result.Fail) =>
-          logger.warn(s"Query failed for server $server, not retrying. Exception was: $e", e)
-          // Wrap SolrException with solrs RemoteSolrException
-          val ex = if(e.isInstanceOf[SolrException]) new RemoteSolrException(500, e.getMessage, e) else e
-          Future.failed(ex)
+    
+    val result = futureFactory.newPromise[QueryResponse]
+    for {
+      queryDone <- inner_query(server, queryContext.q)
+    } yield {
+      queryDone match {
```

Comment:
It is not very nice, sorry, it was just my first thougth.
The real problem here is that the functions returns in this case are not easly composable, you can solve the problem with the primitives given by the future
i.e. in scala

	def recover[U >: T](pf: PartialFunction[Throwable, U])(implicit executor: ExecutionContext): Future[U] = {
	val p = Promise[U]()
	onComplete { v => p complete (v recover pf) }
	p.future
	}

Probably the best solution is to have a single "configurable method" that can do the retries and in case will return on first failure...

Unfortunately my knowledge of Solr are weak and so I'm open to changes!